core-primitives/stf-state-handler/src/file_io.rs

```diff
@@ -19,23 +19,120 @@
 use crate::sgx_reexport_prelude::*;
 
 #[cfg(any(test, feature = "std"))]
-use rust_base58::base58::ToBase58;
+use rust_base58::base58::{FromBase58, ToBase58};
 
 #[cfg(feature = "sgx")]
-use base58::ToBase58;
-
-#[cfg(any(test, feature = "sgx"))]
-use itp_settings::files::ENCRYPTED_STATE_FILE;
+use base58::{FromBase58, ToBase58};
 
 #[cfg(any(test, feature = "sgx"))]
 use std::string::String;
 
-use crate::{error::Result, state_snapshot_primitives::StateId};
-use codec::Encode;
+use crate::{
+	error::{Error, Result},
+	state_snapshot_primitives::StateId,
+};
+use codec::{Decode, Encode};
+// Todo: Can be migrated to here in the course of #1292.
 use itp_settings::files::SHARDS_PATH;
 use itp_types::ShardIdentifier;
 use log::error;
-use std::{format, path::PathBuf, vec::Vec};
+use std::{
+	format,
+	path::{Path, PathBuf},
+	vec::Vec,
+};
+
+/// File name of the encrypted state file.
+///
+/// It is also the suffix of all past snapshots.
+pub const ENCRYPTED_STATE_FILE: &str = "state.bin";
+
+/// Helps with file system operations of all files relevant for the State.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct StateDir {
+	base_path: PathBuf,
+}
+
+impl StateDir {
+	pub fn new(base_path: PathBuf) -> Self {
+		Self { base_path }
+	}
+
+	pub fn shards_directory(&self) -> PathBuf {
+		self.base_path.join(SHARDS_PATH)
+	}
+
+	pub fn shard_path(&self, shard: &ShardIdentifier) -> PathBuf {
+		self.shards_directory().join(shard.encode().to_base58())
+	}
+
+	/// Lists any valid shards that are found in the shard path.
+	///
+	/// Ignores any items (files, directories) that are not valid shard identifiers.
+	pub fn list_shards(&self) -> Result<Vec<ShardIdentifier>> {
+		list_shards(&self.shards_directory())
+	}
+
+	/// Lists all files with a valid state snapshot naming pattern.
+	pub fn list_state_ids_for_shard(
+		&self,
+		shard_identifier: &ShardIdentifier,
+	) -> Result<Vec<StateId>> {
+		let shard_path = self.shard_path(shard_identifier);
+		let directory_items = list_items_in_directory(&shard_path);
+
+		Ok(directory_items
+			.iter()
+			.filter_map(|item| {
+				let maybe_state_id = extract_state_id_from_file_name(item.as_str());
+				if maybe_state_id.is_none() {
+					log::warn!("Found item ({}) that does not match state snapshot naming pattern, ignoring it", item)
+				}
+				maybe_state_id
+			})
+			.collect())
+	}
+
+	pub fn purge_shard_dir(&self, shard: &ShardIdentifier) {
+		let shard_dir_path = self.shard_path(shard);
+		if let Err(e) = std::fs::remove_dir_all(&shard_dir_path) {
+			error!("Failed to remove shard directory {:?}: {:?}", shard_dir_path, e);
+		}
+	}
+
+	pub fn shard_exists(&self, shard: &ShardIdentifier) -> bool {
+		let shard_path = self.shard_path(shard);
+		if !shard_path.exists() {
+			return false
+		}
+
+		shard_path
+			.read_dir()
+			// When the iterator over all files in the directory returns none, the directory is empty.
+			.map(|mut d| d.next().is_some())
+			.unwrap_or(false)
```

Comment:
How about this?
```suggestion
    shard_path.exists() &&
    // When the iterator over all files in the directory returns none, the directory is empty.
    shard_path.read_dir()
        .map(|dir_iter| dir_iter.next().is_some())
        .unwrap_or(false)

```

The short circuiting will happen automatically in this case ;)

Reply:
You made me go down the rabbit hole of looking at all the function implementations from before, and I refactored and extracted functions in quite some places. XD

In this particular case, we now have:

```rust
shard_path().exitsts() && shard_contains_valid_state_id(&shard_path)
```

Which also changes the semantics a bit to be more restrictive. Before we just checked if any file is present, now we check if there is at least a valid state snapshot file, which, I think, is better.

However, if I had done this with the previous state of the code, we would have only worked with `Vec`s, which meant that we decoded all files into valid state ids and then check if the `Vec`'s length is at least 1. So I wanted to change this into an iterator and return `true` for the first element that is a valid state id. However, while doing this, I also noticed that we collect iterators in various places just to turn them into iterators again; hence I refactored some more code to return iterators into vecs. I like these optimization tasks. :)